File: grokkingfp-in-clojure/src/lagenorhynque/ch03_abbreviate_names.clj
(ns lagenorhynque.ch03-abbreviate-names
  "第3章 イミュータブルな値"
  (:require
   [clojure.spec.alpha :as s]
   [clojure.string :as str]))

;; 3.15 コーヒーブレイク：イミュータブルなString API

(s/fdef abbreviate
  :args (s/cat :name string?)
  :ret string?)

(defn abbreviate [name]
  (let [[first-name last-name] (str/split name #" ")
        initial (first first-name)]
    (str initial ". " last-name)))

File: grokkingfp-in-haskell/src/Ch03_AbbreviateNames.hs
{- |
第3章 イミュータブルな値
-}
module Ch03_AbbreviateNames(module Ch03_AbbreviateNames) where

-- 3.15 コーヒーブレイク：イミュータブルなString API

abbreviate :: String -> String
abbreviate name = initial : ". " ++ lastName
  where
    (firstName:lastName:_) = words name
    initial = head firstName
